examples/classic/distributed/TwoPhaseCommit/two_phase_commit.qnt

```diff
@@ -0,0 +1,162 @@
+// -*- mode: Bluespec; -*-
+
+/* This specification stems from the specification of the two phase protocol
+   by Leslie Lamport:
+
+   https://github.com/tlaplus/Examples/blob/master/specifications/transaction_commit/TwoPhase.tla */
+
+module two_phase_commit {
+  type ResourceManagerState = Working | Prepared | Committed | Aborted
+  var resourceManagerStates: str -> ResourceManagerState
+
+  type TransactionManagerState = Init | Done
+  var transactionManagerState: TransactionManagerState
+
+  /// The set of resource managers from which the transaction manager has
+  /// received RMPrepared messages
+  var preparedResourceManagers: Set[str]
+
+  type Message =
+    | Abort
+    | Commit
+    | RMPrepared(str)
+    | RMCommitted(str)
+
+  var messages: Set[Message]
+
+  const resourceManagers: Set[str]
+
+  action init = all {
+    resourceManagerStates' = resourceManagers.mapBy(r => Working),
+    transactionManagerState' = Init,
+    preparedResourceManagers' = Set(),
+    messages' = Set(),
+  }
+
+  /* Transaction Manager actions */
+
+  /// The transaction manager receives a RMPrepared(rm) message from the resource manager rm
+  action addPreparedResourceManager(resourceManager: str): bool = all {
+    transactionManagerState == Init,
+    messages.contains(RMPrepared(resourceManager)),
+    preparedResourceManagers' = preparedResourceManagers.union(Set(resourceManager))
+  }
+
+  /// The transaction manager commits the transaction; enabled iff the
+  /// transaction manager is in its initial state and every resource manager has
+  /// sent an RMPrepared message.
+  action decideOnCommit = all {
+    transactionManagerState == Init,
+    preparedResourceManagers == resourceManagers,
+    transactionManagerState' = Done,
+    messages' = messages.union(Set(Commit))
+  }
+
+  /// The transaction manager spontaneously aborts the transaction
+  action decideOnAbort = all {
+    transactionManagerState == Init,
+    transactionManagerState' = Done,
+    messages' = messages.union(Set(Abort))
+  }
+
+  /* Resource Manager actions */
+
+  /// A resource manager prepares
+  action prepare(resourceManager: str): bool = all {
+    resourceManagerStates.get(resourceManager) == Working,
+    resourceManagerStates' = resourceManagerStates.set(resourceManager, Prepared),
+    messages' = messages.union(Set(RMPrepared(resourceManager)))
+  }
+
+  /// A resource manager aborts, either spontaneously or because of an Abort message
+  action abort(resourceManager: str): bool = all {
+    or {
+      // Resource manager chooses to abort
+      resourceManagerStates.get(resourceManager) == Working,
+      // Resource manager receives an abort message from the transaction manager
+      messages.contains(Abort),
+    },
+    resourceManagerStates' = resourceManagerStates.set(resourceManager, Aborted),
+    // There is no need to let others know that this resource manager aborted
+    messages' = messages
```

Comment:
Why add the unchanged update here, just for this variable, if not including it in the other actions?

Reply:
The motivation is to be able to group this with similar actions, `prepare` and `commit`, in the `step` definition - see line 93-98.

I guess it is a bit weird comparing to the other actions. But not if you think in terms of the effects for those actions - I don't want the amount of unchanged to be the same, I just want the update effect to be the same, so I can combine them in an `any`. With effect annotations, it would be clear that the effect is the same between `prepare`, `commit` and `abort` - and the unchanged part would be invisible since we'd infer it from the annotation.

Reply:
This makes sense to me, esp. after our conversation discussing ways of thinking about partitions of the state space.